src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.GrainDirectory;
+using Orleans.SystemTargetInterfaces;
+
+namespace Orleans.Runtime.GrainDirectory 
+{
+    /// <summary>
+    /// A class that encapsulates response processing logic.
+    /// It is a promise that fires once it has enough responses to make a determination.
+    /// </summary>
+    internal class GlobalSingleInstanceResponseTracker : TaskCompletionSource<GlobalSingleInstanceResponseTracker.Outcome> {
```

Comment:
The more I look at this tracker class and its usages, the more I feel we can improve it (change the approach so it's more maintainable). It has too many moving parts, not to mention that it relies on using the array as mutable state, and then other places (such as the `GlobalSingleInstanceRegistrar.SendRequest`) accesses that array and mutates values by index, and then tell the tracker to re-process...
For now I'm leaving it until I finish reviewing the rest, but I'll try to take a stab at updating it later.